### PR TITLE
Enhancement/3167 implement selector all properties

### PR DESCRIPTION
**Pull request by danielgent**
## Summary

Addresses issue #3167

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/modules/analytics/datastore/__fixtures__/properties-ga4.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
+[
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/__fixtures__/properties-ga4.json:1]: Copied from https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics-4/datastore/__fixtures__/properties.json

File: assets/js/modules/analytics/datastore/__snapshots__/properties.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,668 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/__snapshots__/properties.test.js.snap:1]: I'm not a huge fan of snapshots but this seemed the easiest way of checking the sort order!

File: assets/js/modules/analytics/datastore/properties.js [hunk 3]
```diff
@@ -407,6 +408,33 @@ const baseSelectors = {
 
 		return select( STORE_NAME ).isFetchingGetPropertiesProfiles( accountID );
 	} ),
+
+	/**
+	 * Gets all "classic" Universal Analytic and Google Analytics 4 (GA4) properties this account can access.
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/properties.js:413]: :eyes: my terminology here. I took this from the GA4 Google docs we have

File: assets/js/modules/analytics/datastore/properties.js [hunk 4]
```diff
@@ -407,6 +408,33 @@ const baseSelectors = {
...
+	getPropertiesIncludingGA4: createRegistrySelector( ( select ) => ( state, accountID ) => {
+		const properties = select( STORE_NAME ).getProperties( accountID );
+
+		let propertiesGA4 = [];
+
+		if ( select( 'modules/analytics-4' ) ) {
+			propertiesGA4 = select( 'modules/analytics-4' ).getProperties( accountID );
+		}
+
+		const propertiesGA4Mapped = propertiesGA4.map( ( property ) => ( { ...property, id: property._id } ) );
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/properties.js:434]: according to the fixtures (assuming these to be correct) `ua` properties have an `id` key, but `ga4` have an `_id` key

File: assets/js/modules/analytics/datastore/properties.test.js [hunk 5]
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
...
+
+				// Load data into stores
+				registry.dispatch( STORE_NAME ).receiveGetProperties( fixtures.propertiesProfiles.properties, { accountID } );
+				registry.dispatch( 'modules/analytics-4' ).receiveGetProperties( fixtures.propertiesGA4, { accountID } );
+
+				const properties = registry.select( STORE_NAME ).getPropertiesIncludingGA4( testAccountID );
+
+				expect( properties ).toMatchSnapshot();
+				expect( properties ).toHaveLength( 23 );
+			} );
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/properties.test.js:271]: I would also like to test for when the `modules/analytics-4` datastore is not available, but I don't any way to conditionally load data stores when testing

the test util registerAllStoresOn registers all the stores!
https://github.com/google/site-kit-wp/blob/4cac32bb3a3a5fc977d07546ef3fe5d7f3bc94c3/tests/js/utils.js#L359

File: assets/js/modules/analytics/datastore/properties.test.js [hunk 6]
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
...
 			} );
 		} );
+		describe( 'getPropertiesIncludingGA4', () => {
+			it( 'returns a sorted list of ua and ga4 properties ', async () => {
+				const testAccountID = fixtures.profiles[ 0 ].accountId; // eslint-disable-line sitekit/acronym-case
+				const accountID = testAccountID;
+
+				// Load data into stores
+				registry.dispatch( STORE_NAME ).receiveGetProperties( fixtures.propertiesProfiles.properties, { accountID } );
+				registry.dispatch( 'modules/analytics-4' ).receiveGetProperties( fixtures.propertiesGA4, { accountID } );
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/properties.test.js:265]: Copied store hydration code from previous tests

I could also test for when both of these selectors have not yet resolved *but* that feels out of scope for this as we are already testing that in the respective selector's tests

File: assets/js/modules/analytics/datastore/properties.js [hunk 7]
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
...
+	 * @param {Object} state     Data store's state.
+	 * @param {string} accountID The Analytics Account ID to fetch properties for.
+	 * @return {(Array.<Object>|undefined)} An array of Analytics properties; `undefined` if not loaded.
+	 */
+	getPropertiesIncludingGA4: createRegistrySelector( ( select ) => ( state, accountID ) => {
+		let properties = select( STORE_NAME ).getProperties( accountID );
+
+		if ( select( MODULES_ANALYTICS_4 ) ) {
+			const	propertiesGA4 = select( MODULES_ANALYTICS_4 ).getProperties( accountID );
+			const propertiesGA4Mapped = propertiesGA4.map( ( property ) => ( { ...property, id: property._id } ) );
```
Comment (danielgent) [inline assets/js/modules/analytics/datastore/properties.js:431]: according to the fixtures (assuming these to be correct) `ua` properties have an `id` key, but `ga4` have an `_id` key

File: assets/js/modules/analytics/datastore/properties.js [hunk 8]
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
...
+			properties = properties.concat( propertiesGA4Mapped );
+		}
+
+		return sortBy( properties, [ 'name', 'id' ] );
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.js:435]: @danielgent i think it won't work as expected because GA4 properties store property names in the `displayName` attribute. Let's rewrite this to use custom compare function `properties.sort( ( a, b ) => { ... } )`.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @danielgent. It's a good start. Found a few issues that need to be addressed before we can merge this PR. Could you please take a look at it?

File: assets/js/modules/analytics/datastore/properties.js [hunk 9]
```diff
@@ -407,6 +409,31 @@ const baseSelectors = {
...
+	 *
+	 * @param {Object} state     Data store's state.
+	 * @param {string} accountID The Analytics Account ID to fetch properties for.
+	 * @return {(Array.<Object>|undefined)} An array of Analytics properties; `undefined` if not loaded.
+	 */
+	getPropertiesIncludingGA4: createRegistrySelector( ( select ) => ( state, accountID ) => {
+		let properties = select( STORE_NAME ).getProperties( accountID );
+
+		if ( select( MODULES_ANALYTICS_4 ) ) {
+			const	propertiesGA4 = select( MODULES_ANALYTICS_4 ).getProperties( accountID );
```
[the remark this answers is not in the document]
Comment (danielgent) [reply assets/js/modules/analytics/datastore/properties.js:430]: @eugene-manuilov I increased the size of tabs in vs-code so I can spot these tabs
tabs-spaces

I also found a setting where vscode was rewriting tabs as spaces :facepalm: 
`"editor.insertSpaces": false,`

So hopefully I'll make less of these formatting issues now

File: assets/js/modules/analytics/datastore/properties.js (hunk 8 above)
Comment (danielgent) [reply assets/js/modules/analytics/datastore/properties.js:435]: @eugene-manuilov so I can differentiate between the `ga` and `ua4` properties based on if they have the key `_id` or `id`, right? (from looking at the fixtures)

So I can do that.  :+1:   (and drop the `propertiesGA4Mapped` above)

File: assets/js/modules/analytics/datastore/properties.test.js [hunk 10]
```diff
@@ -255,6 +255,22 @@ describe( 'modules/analytics properties', () => {
...
+
+				const properties = registry.select( STORE_NAME ).getPropertiesIncludingGA4( testAccountID );
+
+				expect( properties ).toMatchSnapshot();
```
[the remark this answers is not in the document]
Comment (danielgent) [reply assets/js/modules/analytics/datastore/properties.test.js:270]: @eugene-manuilov  Makes total sense!  I can assert that each element is in the correct place, and then anyone else can quickly check that I've done the ticket correctly

**Comment by danielgent** [comment]
@eugene-manuilov  I think I've addressed all your feedback now! Thanks for that.

There are quite a few changes so I imagine this will require another round of feedback.  All suggestions welcome :smile:

File: assets/js/modules/analytics/datastore/properties.test.js [hunk 11]
```diff
@@ -255,6 +259,98 @@ describe( 'modules/analytics properties', () => {
...
+				expect( properties[ 2 ].id ).toBe( 'UA-151753095-1' );
+				expect( properties[ 3 ]._id ).toBe( '151753095-3' );
+
+				expect( properties[ 0 ].name ).toBe( 'rwh' );
+				expect( properties[ 1 ].displayName ).toBe( 'troubled-tipped.example.com' );
+				expect( properties[ 2 ].name ).toBe( 'troubled-tipped.example.com' );
+				expect( properties[ 3 ].displayName ).toBe( 'www.elasticpress.io' );
+			} );
+
+			it( 'returns only ua properties when ga4 datastore not available ', async () => {
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.test.js:314]: @danielgent please take my apologies, but I think we don't need this test because it does nothing but repeats the test above. Let's remove this.

File: assets/js/modules/analytics/datastore/properties.js [hunk 12]
```diff
@@ -325,6 +326,18 @@ const baseResolvers = {
...
+
+function compare( a, b ) {
+	if ( a < b ) {
+		return -1;
+	}
+	if ( a > b ) {
+		return 1;
+	}
+	return 0;
+}
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.js:339]: These functions are very specific to the `getPropertiesIncludingGA4` selector and used only there. Let's move it into the selector body to encapsulate it there.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Good work, @danielgent. There are two minor things to address and I'll approve this PR. Could you please update it?

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Good work, @danielgent. There are two minor things to address and I'll approve this PR. Could you please update it?

File: assets/js/modules/analytics/datastore/properties.js (hunk 12 above)
Comment (danielgent) [reply assets/js/modules/analytics/datastore/properties.js:339]: Sure thing @eugene-manuilov !  I was half expecting you to point out that they already existed in native JS or in Lodash :see_no_evil:

File: assets/js/modules/analytics/datastore/properties.test.js (hunk 11 above)
Comment (danielgent) [reply assets/js/modules/analytics/datastore/properties.test.js]: No problem @eugene-manuilov 

It just tests that the code I've written doesn't explode if the ga4 datastore isn't available.  It's not where I'd expect it to fail though :+1:

File: assets/js/modules/analytics/datastore/properties.js [hunk 13]
```diff
@@ -407,6 +408,54 @@ const baseSelectors = {
...
+
+		function compare( a, b ) {
+			if ( a < b ) {
+				return -1;
+			}
+			if ( a > b ) {
+				return 1;
+			}
+			return 0;
+		}
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.js:443]:
```suggestion
		const isGA4 = ( property ) => !! property._id;
		const compare = ( a, b ) => {
			if ( a < b ) {
				return -1;
			}
			if ( a > b ) {
				return 1;
			}
			return 0;
		}
```

File: assets/js/modules/analytics/datastore/properties.test.js [hunk 14]
```diff
@@ -255,6 +256,59 @@ describe( 'modules/analytics properties', () => {
...
+						_id: '151753095-3',
+						_accountID: '151753095',
+						displayName: 'www.elasticpress.io',
+					},
+					{
+						_id: '151753095-4',
+						_accountID: '151753095',
+						displayName: 'troubled-tipped.example.com',
+					},
+				], { accountID } );
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.test.js:294]:
```suggestion

				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetProperties(
					[
						{
							_id: '151753095-3',
							_accountID: '151753095',
							displayName: 'www.elasticpress.io',
						},
						{
							_id: '151753095-4',
							_accountID: '151753095',
							displayName: 'troubled-tipped.example.com',
						},
					],
					{ accountID }
				);
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Awesome! Thanks, @danielgent! Looks good to me. There are two formatting issues, I'll take care of it.

File: assets/js/modules/analytics/datastore/properties.js [hunk 15]
```diff
@@ -407,6 +408,53 @@ const baseSelectors = {
...
+		const isGA4 = ( property ) => !! property._id;
+		const compare = ( a, b ) => {
+			if ( a < b ) {
+				return -1;
+			}
+			if ( a > b ) {
+				return 1;
+			}
+			return 0;
+		}
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics/datastore/properties.js:442]: Forgot to update here as well:
```suggestion
		};
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]

**Review by eugene-manuilov — APPROVED** [review]
